Pilz Sequence Motion MoveGroupSequence example (#917)
moveit/moveit2_tutorials: doc/how_to_guides/pilz_industrial_motion_planner/pilz_industrial_motion_planner.rst

   ``blend_radius``\ (i) + ``blend_radius``\ (i+1) has to be smaller
   than the distance between the goals.

Service interface
~~~~~~~~~~~~~~~~~

The service ``plan_sequence_path`` allows the user to generate a joint
trajectory for a ``moveit_msgs::msg::MotionSequenceRequest``.
The trajectory is returned and not executed.

To use the ``MoveGroupSequenceService`` and the ``MoveGroupSequenceAction``, refer to the :codedir:`Pilz Motion Planner sequence example <how_to_guides/pilz_industrial_motion_planner/src/pilz_sequence.cpp>`.
To run this, execute the following commands in separate Terminals:

::

    ros2 launch moveit2_tutorials pilz_moveit.launch.py
    ros2 run moveit2_tutorials pilz_sequence

> For this service and action, the move_group launch file needs to be modify to include these Pilz Motion Planner capabilities.

Reviewer: ```suggestion
For this service and action, the move_group launch file needs to be modified to include these Pilz Motion Planner capabilities.

       .robot_description(file_path="config/panda.urdf.xacro")
       .trajectory_execution(file_path="config/gripper_moveit_controllers.yaml")
       .planning_scene_monitor(
          publish_robot_description=True, publish_robot_description_semantic=True
       )
       .planning_pipelines(
          pipelines=["pilz_industrial_motion_planner"]
       )
       .to_moveit_configs()
    )

    # Starts Pilz Industrial Motion Planner MoveGroupSequenceAction and MoveGroupSequenceService servers
    move_group_capabilities = {
       "capabilities": "pilz_industrial_motion_planner/MoveGroupSequenceAction pilz_industrial_motion_planner/MoveGroupSequenceService"
    }

The
:codedir:`pilz_sequence.cpp file <how_to_guides/pilz_industrial_motion_planner/src/pilz_sequence.launch.py>`
creates two target poses that will be reached sequentially.

::

    // ----- Motion Sequence Item 1
    // Create a MotionSequenceItem
    moveit_msgs::msg::MotionSequenceItem item1;

    // Set pose blend radius
    item1.blend_radius = 0.1;

    // MotionSequenceItem configuration
    item1.req.group_name = PLANNING_GROUP;
    item1.req.planner_id = "LIN";

>     item1.req.allowed_planning_time = 5;

Reviewer: ```suggestion
    item1.req.allowed_planning_time = 5.0;

    item1.req.max_velocity_scaling_factor = 0.1;
    item1.req.max_acceleration_scaling_factor = 0.1;

    moveit_msgs::msg::Constraints constraints_item1;
    moveit_msgs::msg::PositionConstraint pos_constraint_item1;
    pos_constraint_item1.header.frame_id = "world";
    pos_constraint_item1.link_name = "panda_hand";

    // Set a constraint pose
    auto target_pose_item1 = [] {
       geometry_msgs::msg::PoseStamped msg;
       msg.header.frame_id = "world";
       msg.pose.position.x = 0.3;
       msg.pose.position.y = -0.2;
       msg.pose.position.z = 0.6;
       msg.pose.orientation.x = 1.0;
       msg.pose.orientation.y = 0;
       msg.pose.orientation.z = 0;

>        msg.pose.orientation.w = 0;

Reviewer: ```suggestion
       msg.pose.orientation.y = 0.0;
       msg.pose.orientation.z = 0.0;
       msg.pose.orientation.w = 0.0;

       return msg;
    } ();
    item1.req.goal_constraints.push_back(kinematic_constraints::constructGoalConstraints("panda_link8", target_pose_item1));

The service client needs to be initialized:

::

    // MoveGroupSequence service client
    using GetMotionSequence = moveit_msgs::srv::GetMotionSequence;
    auto service_client = node->create_client<GetMotionSequence>("/plan_sequence_path");

    // Verify that the action server is up and running
    while (!service_client->wait_for_service(std::chrono::seconds(10)))
    {
       RCLCPP_WARN(LOGGER, "Waiting for service /plan_sequence_path to be available...");
    }

Then, the request is created:

::

    // Create request
    auto service_request = std::make_shared<GetMotionSequence::Request>();
    service_request->request.items.push_back(item1);
    service_request->request.items.push_back(item2);

> Service call and response. The method ``future.wait_for(timeout_duration)`` waits for the

Reviewer: This "Service call and response" is not a full sentence, could you reword?

Author: I changed the whole sentence to "Once the service call is completed, the method ``future.wait_for(timeout_duration)`` blocks until a specified ``timeout_duration`` has elapsed or the result becomes available, whichever comes first. The return value identifies the state of the result. This is performed every second until the future is ready."

result to become available. Blocks until specified ``timeout_duration`` has elapsed or the
result becomes available, whichever comes first. The return value identifies the state of

> the result. This is perform every 1 seconds until the future is ready.

Reviewer: ```suggestion
the result. This is performed every 1 second until the future is ready.

To run this, execute the following commands in separate Terminals:

::

    ros2 launch moveit2_tutorials pilz_moveit.launch.py
    ros2 run moveit2_tutorials pilz_sequence

For this action and service, the move_group launch file needs to be modify to include these Pilz Motion Planner capabilities.
The new 
:codedir:`the pilz_moveit.launch.py <how_to_guides/pilz_industrial_motion_planner/launch/pilz_moveit.launch.py>` 
is used instead:

::

    moveit_config = (
       MoveItConfigsBuilder("moveit_resources_panda")
       .robot_description(file_path="config/panda.urdf.xacro")
       .trajectory_execution(file_path="config/gripper_moveit_controllers.yaml")
       .planning_scene_monitor(
          publish_robot_description=True, publish_robot_description_semantic=True
       )
       .planning_pipelines(
          pipelines=["pilz_industrial_motion_planner"]
       )
       .to_moveit_configs()
    )
 
    # Starts Pilz Industrial Motion Planner MoveGroupSequenceAction and MoveGroupSequenceService servers
    move_group_capabilities = {
       "capabilities": "pilz_industrial_motion_planner/MoveGroupSequenceAction pilz_industrial_motion_planner/MoveGroupSequenceService"
    }

> The sequence script creates 2 targets poses that will be reach sequentially.

Reviewer: ```suggestion
The `pilz_sequence.cpp file` creates 2 target poses that will be reached sequentially.
```

... and see if you can directly link to the source code with the `:codedir:` directive.

    // Create a MotionSequenceItem
    moveit_msgs::msg::MotionSequenceItem item1;
    
    // Set pose blend radius
    item1.blend_radius = 0.1;
 
    // MotionSequenceItem configuration
    item1.req.group_name = PLANNING_GROUP;
    item1.req.planner_id = "LIN";
    item1.req.allowed_planning_time = 5;
    item1.req.max_velocity_scaling_factor = 0.1;
    item1.req.max_acceleration_scaling_factor = 0.1;
          
    moveit_msgs::msg::Constraints constraints_item1;
    moveit_msgs::msg::PositionConstraint pos_constraint_item1;
    pos_constraint_item1.header.frame_id = "world";
    pos_constraint_item1.link_name = "panda_hand";
    
    // Set a constraint pose
    auto target_pose_item1 = [] {
       geometry_msgs::msg::PoseStamped msg;
       msg.header.frame_id = "world";
       msg.pose.position.x = 0.3;
       msg.pose.position.y = -0.2;
       msg.pose.position.z = 0.6;
       msg.pose.orientation.x = 1.0;
       msg.pose.orientation.y = 0;
       msg.pose.orientation.z = 0;
       msg.pose.orientation.w = 0;
       return msg;
    } ();
    item1.req.goal_constraints.push_back(kinematic_constraints::constructGoalConstraints("panda_link8", target_pose_item1));

> The action client needs to be initialize:

Reviewer: ```suggestion
The action client needs to be initialized:

    sequence_request.items.push_back(item1);
    sequence_request.items.push_back(item2);

Create goal and planning options. A goal response callback and result callback can be included as well.

::

    // Create action goal
    auto goal_msg = MoveGroupSequence::Goal();
    goal_msg.request = sequence_request;
 
    // Planning options
    goal_msg.planning_options.planning_scene_diff.is_diff = true;
    goal_msg.planning_options.planning_scene_diff.robot_state.is_diff = true;
    // goal_msg.planning_options.plan_only = true;

Finally, send the goal request and wait for the response:

::

    // Send the action goal
    auto goal_handle_future = client->async_send_goal(goal_msg, send_goal_options);
 
    // Get result
    auto future_result = client->async_get_result(goal_handle_future.get());
 
    // Wait for the result
    if (future_result.valid()) {
       auto result = future_result.get();  // Blocks the program execution until it gets a response
       RCLCPP_INFO(LOGGER, "Action completed. Result: %d",  static_cast<int>(result.code));
    } else {
       RCLCPP_ERROR(LOGGER, "Action couldn't be completed.");
    }

> If the motion need to be stopped mid trajectory, the action have to be cancel with:

Reviewer: ```suggestion
If the motion needs to be stopped mid-execution, the action can be canceled with:

::

    auto future_cancel_motion = client->async_cancel_goal(goal_handle_future_new.get());

The ``pilz_robot_programming`` package for a `ROS 1 Python script
<https://github.com/PilzDE/pilz_industrial_motion/blob/melodic-devel/pilz_robot_programming/examples/demo_program.py>`_

> shows how to use the capability.

Reviewer: we can probably just remove this paragraph, unless you still find it helpful?

Author: The basics are all covered I think, I just left it because it has some methods that are not in ROS2 yet, like the **pause** command. Nevertheless these commands are part of the pilz_robot_programming, not the Pilz Industrial Motion Planner.

I will delete the paragraph to not create confusion.

Service interface
~~~~~~~~~~~~~~~~~

The service ``plan_sequence_path`` allows the user to generate a joint
trajectory for a ``moveit_msgs::msg::MotionSequenceRequest``.
The trajectory is returned and not executed.

> The service client needs to be initialize:

Reviewer: ```suggestion
The service client needs to be initialized:

::

    // MoveGroupSequence service client
    using GetMotionSequence = moveit_msgs::srv::GetMotionSequence;
    auto service_client = node->create_client<GetMotionSequence>("/plan_sequence_path");
 
    // Verify that the action server is up and running
    while (!service_client->wait_for_service(std::chrono::seconds(10))) {
       RCLCPP_WARN(LOGGER, "Waiting for service /plan_sequence_path to be available...");
    }

Then, the request is created:

::

    // Create request
    auto service_request = std::make_shared<GetMotionSequence::Request>();
    service_request->request.items.push_back(item1);
    service_request->request.items.push_back(item2);

> Service call and response. The method ``future.get()`` blocks the execution of the program until the server response arrives.

Reviewer: Actually, `future.get()` will not block and wait for a response, which is maybe what you were running into!

You should be doing something like this:
https://docs.ros.org/en/humble/Tutorials/Beginner-Client-Libraries/Writing-A-Simple-Cpp-Service-And-Client.html#write-the-client-node

Author: That was just a comment! It is not a issue, but it is how the service works, as long as I've seen. In the Pilz Industrial Motion Planner package, the class MoveGroupSequenceService of the [move_group_sequence_service.h](https://github.com/moveit/moveit2/blob/main/moveit_planners/pilz_industrial_motion_planner/include/pilz_industrial_motion_planner/move_group_sequence_service.h) only has a **plan** method. On the other hand, the MoveGroupSequenceAction of the [move_group_sequence_action.h](https://github.com/moveit/moveit2/blob/main/moveit_planners/pilz_industrial_motion_planner/include/pilz_industrial_motion_planner/move_group_sequence_action.h) has plan and execution methods.

I didn´t use the rclcpp::spin_until_future_complete since the node is already spinning, set by the executor at the beginning of the code. That gave me some error when compiling. However, this is just necessary for the visuals, so we could forget about the trajectory representation and do it the right way.

Reviewer 2: If the node is already spinning elsewhere, then you likely just want to do a `future.wait_for(1s)` or something like that, and then check its status:

https://en.cppreference.com/w/cpp/thread/future/wait_for

See, for example, line 152 of [this link](https://docs.ros2.org/latest/api/rclcpp/static__single__threaded__executor_8hpp_source.html), but then use a non-zero timeout.

Reviewer: If the node is already spinning elsewhere, then you likely just want to do a `future.wait_for(1s)` or something like that, and then check its status:

https://en.cppreference.com/w/cpp/thread/future/wait_for

See, for example, line 152 of [this link](https://docs.ros2.org/latest/api/rclcpp/static__single__threaded__executor_8hpp_source.html), but then use a non-zero timeout.

Author: I tried it and it works, but if the timeout is too low it doesn´t work, since it returns before the actions (or service) is finished. The [wait_for documentation](https://en.cppreference.com/w/cpp/thread/shared_future/wait_for) shows an example using a while loop to use multiple future.wait_for(1s) until the goal is reached. 

```
do {
  switch (action_status = future_result.wait_for(std::chrono::seconds(1)); action_status)
  {
    case std::future_status::deferred:
      std::cout << "Deferred\n";
      break;
    case std::future_status::timeout:
      std::cout << "Executing trajectory...\n";
      break;
    case std::future_status::ready:
      std::cout << "Action ready!\n";
      break;
  }    
}
while (action_status != std::future_status::ready);
```
This works fine and we can have the RVIZ visuals at the same time. I should change the std::cout with the log.

In the tutorial I could write how to use the rclcpp::spin_until_future_complete as well.

::

    // Call the service and process the result
    auto future = service_client->async_send_request(service_request);
 
    // Function to draw the trajectory
    auto const draw_trajectory_tool_path =
       [&moveit_visual_tools, jmg = move_group_interface.getRobotModel()->getJointModelGroup("panda_arm")](
          auto const& trajectories) {
       for (const auto& trajectory : trajectories) {
          moveit_visual_tools.publishTrajectoryLine(trajectory, jmg);
       }
    };
 
    auto response = future.get();
    if (response->response.error_code.val == moveit_msgs::msg::MoveItErrorCodes::SUCCESS) {
       RCLCPP_INFO(LOGGER, "Planning successful");
 
       // Access the planned trajectory
       auto trajectory = response->response.planned_trajectories;
       draw_trajectory_tool_path(trajectory);
       moveit_visual_tools.trigger();
 
    } else {
       RCLCPP_ERROR(LOGGER, "Planning failed with error code: %d", response->response.error_code.val);
    }

> In this case, the planned trajectory is drawn.

Reviewer: Consider adding a screenshot here!

already satisfied but the ``MoveGroupSequenceAction`` capability doesn't
implement such a check to allow moving on a circular or comparable path.

The action client needs to be initialized:

::

    // MoveGroupSequence action client
    using MoveGroupSequence = moveit_msgs::action::MoveGroupSequence;
    auto action_client = rclcpp_action::create_client<MoveGroupSequence>(node, "/sequence_move_group");

    // Verify that the action server is up and running
    if (!action_client->wait_for_action_server(std::chrono::seconds(10)))
    {
       RCLCPP_ERROR(LOGGER, "Error waiting for action server /sequence_move_group");
       return -1;
    }

Then, the request is created:

::

    // Create a MotionSequenceRequest
    moveit_msgs::msg::MotionSequenceRequest sequence_request;
    sequence_request.items.push_back(item1);
    sequence_request.items.push_back(item2);

> Create goal and planning options. A goal response callback and result callback can be included as well.

Reviewer: Same here: "Create goal and planning options" isn't quite a full sentence, could you reword?

Author: Something simple like "The goal and planning options are configured." ?
